### ISSUE TITLE
Add support for SAM optimizer

### ISSUE THREAD

**Issue by fcakyon** [open-mmlab/mmdetection#4747]
**Add support for [SAM based optimization](https://github.com/davda54/sam)**

**Motivation**
In Deep Learning we use optimization algorithms such as SGD/Adam to achieve convergence in our model, which leads to finding the global minima, i.e a point where the loss of the training dataset is low. But several kinds of research such as [Zhang et al](https://arxiv.org/abs/1611.03530) have shown, many networks can easily memorize the training data and have the capacity to readily overfit, To prevent this problem and add more generalization, Researchers at Google have published a new paper called [Sharpness Awareness Minimization](https://arxiv.org/abs/2010.01412) which provides State of the Art results on CIFAR10 and other datasets.

**Related resources**
Official pytorch based repo: https://github.com/davda54/sam

**Comment by xvjiarui** [comment]
Hi @fcakyon 
We don't have plan for it for now. But PRs are welcome.

**Comment by guillaumefrd** [comment]
I would also be interested in the addition of this feature.
